Add type hints

bbannier opened the pull request:
This PR adds type hints to most functions, and adds a pre-commit hook running mypy to validate typing. I did not attempt to clean up any of the APIs, but having proper typing could be a first step for that. It also makes it easier to modify and extend code safely, and helped identify a few incorrect typings which would otherwise manifest as runtime errors.

awelzel commented on zeekpkg/manager.py:
```diff
@@ -1126,10 +1126,8 @@ def _refresh_source(self, name, aggregate=False, push=False):
-                            url,
-                            error,
+                        LOG.warning(
+                            f"failed to clone {url}, skipping aggregation: {error}",
```
No strong opinion for zkg, but for logging calls, seems it's still recommended to avoid f-strings elsewhere (reddit has enough hits, eventually leading to this)

https://google.github.io/styleguide/pyguide.html#3101-logging

Not sure that matters for zkg, but in certain environments  querying on the uninterpolated string could get all occurrences of a warning, regardless of which url/error was triggered. That stops working with `f-strings` because the non-interpolated string isn't available anymore.

awelzel commented on zeekpkg/manager.py:
```diff
@@ -1378,7 +1369,7 @@ def remove(self, pkg_path):
                     os.unlink(link)
                 except OSError as err:
-                    LOG.warn("cannot remove link for %s", err)
+                    LOG.warning(f"cannot remove link for {err}")
```
...and the f-string pattern becomes inconsistent with `LOG.debug("removing link %s", link)` above (just happened to spot this).

awelzel commented on zeekpkg/_util.py:
```diff
@@ -206,19 +218,20 @@ def git_default_branch(repo):
 
-            return head_ref_name
+                return head_ref_name
+        except Exception:
```
This now potentially catches `AssertExceptions` raised by the `asserts` above (and nothing else?) which feels pretty off to me for Python:
```
In [2]: try:
   ...:     assert 1 == 2
   ...: except Exception:
   ...:     print("caught")
   ...: 
caught
```

I'd just remove the assert tbh, the original code seemed fine.

awelzel replied:
Hmm, are these asserts some sort of mypy hints? Then ignore my comment. This is just new to me.

Still think if there's a more direct way to indicate the expected type like 
`ref = typing.cast(git.Reference, repo.head.ref)` - seems to do the trick here - we should probably prefer that.

I'm just picking up on the potential AssertionError raised coupled with the unfortunate `except Exception` below.

bbannier replied:
Yepp, the assertions guide mypy so it can remove `None` branches.

awelzel commented on zeekpkg/_util.py:
```diff
@@ -206,19 +218,20 @@ def git_default_branch(repo):
+            ref = repo.head.ref
+            assert hasattr(ref, "name")
+            assert isinstance(ref.name, str)
+            if head_ref_name := ref.name:
```
It seems `head_ref_name` is superflous:
```
if ref.name:
    if ref.name.startswith("origin/"):
        return ref.name.split("/", 1)[-1]
    return ref.name
```

awelzel commented on zeekpkg/_util.py:
```diff
@@ -290,7 +305,7 @@ def git_remote_urls(repo):
 
 
-def is_sha1(s):
+def is_sha1(s: str | None) -> bool:
```
Can we force this to just `str`? Seems reasonable expect a caller to do the None check.

awelzel commented on zeekpkg/manager.py:
```diff
@@ -446,7 +451,7 @@ def _write_plugin_magic(self, ipkg):
                         )
 
-    def _read_manifest(self):
+    def _read_manifest(self) -> tuple[str, str, str]:
```
Mind updating the doc string to match this?

awelzel commented on zeekpkg/manager.py:
```diff
@@ -2722,18 +2832,22 @@ def _stage(self, package, version, clone, stage, env=None):
 
         metadata = _get_package_metadata(metadata_parser)
-        metadata, invalid_reason = self._interpolate_package_metadata(metadata, stage)
+        metadata2, invalid_reason = self._interpolate_package_metadata(metadata, stage)
```
Mind calling this `interpolated_metadata` or `metadata_interp` or is that too long?

awelzel commented on zeekpkg/manager.py:
```diff
@@ -687,16 +692,16 @@ def find_builtin_package(self, pkg_path):
         return [ipkg for _, ipkg in sorted(self.installed_pkgs.items())]
 
-    def installed_package_dependencies(self):
+    def installed_package_dependencies(self) -> dict[str, dict[str, str] | None]:
```
Could we make `ipkg.package.dependencies()` to not return `None` and the one case where it does `raise ValueError` or an Assertion instead?

I understand you didn't want to modify the API, but dependencies should either exist or be empty. Even having `None` is just super weird, especially if both callers seem to ignore `None` anyhow.

bbannier replied:
The issue is that validation isn't well encapsulated and many of these functions just take a naked `metadata: dict[str, str]` which leads to functions doing their own ad hoc validation which is not great (`user_vars` just below has the same pattern). TBH, I'd prefer not trying to clean this up right now.

awelzel replied:
yeah, works for me, I guess it's clear that `dict[str, dict[str, str] | None]` for dependencies is a bit strange :-)

bbannier replied:
Agree 100%. I'd argue that's the kind of cruft one gets when developing without type hints ;)

awelzel reviewed (approved):
Approving, but I'm pretty partial here, so if this helps `zkg`, :ship: and :man_shrugging:

bbannier committed:
Bump ruff pre-commit hooks

bbannier committed:
Fully type all functions

This is mainly a mechanical change where I just made previous types
explicit. In certain places I reorganized code so that we can avoid
making attributes optional, or to avoid shadowing values with bindings
of different type.

bbannier committed:
Migrate from deprecated `LOG.warn`

bbannier committed:
Bump to python-3.10

bbannier committed:
Enable and fix ruff 'PL' lints

bbannier committed:
Change `is_sha` to always expect a value
